src/storage/tde.h

```diff
@@ -0,0 +1,203 @@
+ *
+ */
+
+
+/*
+ * tde.hpp -
+ */
+
+#ifndef _TDE_HPP_
+#define _TDE_HPP_
+
+#ident "$Id$"
+
+#include "storage_common.h"
+
+#if defined (SA_MODE)
+#include "boot_sr.h"
+#endif
+
+struct fileio_page;
+typedef fileio_page FILEIO_PAGE;
+struct log_page;
+typedef log_page LOG_PAGE;
+
+/*
+ * Each value is also used to be index of tde_Algorithm_str[].
+ * These must be changed togeter
+ */
+typedef enum
+{
+  TDE_ALGORITHM_NONE = 0,
+  TDE_ALGORITHM_AES = 1,
+  TDE_ALGORITHM_ARIA = 2,
+} TDE_ALGORITHM;
+
+#define TDE_DK_ALGORITHM TDE_ALGORITHM_AES
+
+/* ENCRYPTION AREA */
+#define TDE_DATA_PAGE_ENC_OFFSET    sizeof (FILEIO_PAGE_RESERVED)
+#define TDE_DATA_PAGE_ENC_LENGTH    DB_PAGESIZE
+#define TDE_LOG_PAGE_ENC_OFFSET     sizeof (LOG_HDRPAGE)
+#define TDE_LOG_PAGE_ENC_LENGTH     ((LOG_PAGESIZE) - (TDE_LOG_PAGE_ENC_OFFSET))
+
+/* 128 bit nonce */
+#define TDE_DATA_PAGE_NONCE_LENGTH  16
+#define TDE_LOG_PAGE_NONCE_LENGTH   16
+#define TDE_DK_NONCE_LENGTH         16
+
+/* TDE Keys - 256 bit */
+#define TDE_MASTER_KEY_LENGTH 32
+#define TDE_DATA_KEY_LENGTH   32
+
+typedef struct tde_mk_file_item
+{
+  time_t created_time;		/* If it is -1, it is invalid and avaliable for new key */
+  unsigned char master_key[TDE_MASTER_KEY_LENGTH];
+} TDE_MK_FILE_ITEM;
+
+#define TDE_MK_FILE_CONTENTS_START  CUBRID_MAGIC_MAX_LENGTH
+#define TDE_MK_FILE_ITEM_SIZE       (sizeof (TDE_MK_FILE_ITEM))
+#define TDE_MK_FILE_ITEM_OFFSET(index) \
+  (TDE_MK_FILE_CONTENTS_START + TDE_MK_FILE_ITEM_SIZE * (index))
+#define TDE_MK_FILE_ITEM_INDEX(offset) \
+  (((offset) - TDE_MK_FILE_CONTENTS_START) / TDE_MK_FILE_ITEM_SIZE)
+
+typedef struct tde_data_key_set
+{
+  unsigned char perm_key[TDE_DATA_KEY_LENGTH];
+  unsigned char temp_key[TDE_DATA_KEY_LENGTH];
+  unsigned char log_key[TDE_DATA_KEY_LENGTH];
+} TDE_DATA_KEY_SET;
+
+enum tde_data_key_type
+{
+  TDE_DATA_KEY_TYPE_PERM,
+  TDE_DATA_KEY_TYPE_TEMP,
+  TDE_DATA_KEY_TYPE_LOG
+};
+typedef tde_data_key_type TDE_DATA_KEY_TYPE;
+
+#if 0
+#if defined(CS_MODE)
+#define TDE_HA_SOCK_NAME ".ha_sock"
+#endif /* CS_MODE */
+#endif /* TDE for replication log is disabled */
+
+#if !defined(CS_MODE)
+/*
+ * TDE module
+ *
+ * Note: Now TDE for replication log is disabled,
+ * so CS_MODE version tde_cipher is not needed.
+ */
+typedef struct tde_cipher
+{
+  bool is_loaded;
+  TDE_DATA_KEY_SET data_keys;
+  int64_t temp_write_counter;	/* used as nonce for temp file page, it has to be dealt atomically */
+} TDE_CIPHER;
+
+extern TDE_CIPHER tde_Cipher;
+
+/*
+ * TDE module stores key information with all tha data keys encrypted and master key hashed.
+ */
+typedef struct tde_keyinfo
+{
+  int mk_index;
+  time_t created_time;
+  time_t set_time;
+  unsigned char mk_hash[TDE_MASTER_KEY_LENGTH];
+  unsigned char dk_perm[TDE_DATA_KEY_LENGTH];
+  unsigned char dk_temp[TDE_DATA_KEY_LENGTH];
+  unsigned char dk_log[TDE_DATA_KEY_LENGTH];
+} TDE_KEYINFO;
+
+/* Is log record contains User Data */
+#define LOG_CONTAINS_USER_DATA(rcvindex) \
```

Comment:
it looks that the condition clause is too long to operate fast.
isn't there any else way to check the condition ?

Reply:
This is written with reference to LOG_IS_MVCC_OPERATION() macro, which is also used when creating a log record to check recovery indices more than 10. Log type and log recovery index are independent with each other, so I think all the possible recovery indices should be checked. 

I don't think it can be a serious bottleneck actually because there are more expensive operations around this code like fixing page or copying memory. However, you are right, it is long and ugly. Do you have a better idea to check all possible recovery index?

Reply:
how about defining an array LOG_CONTAINS_USER_DATA as below:

#define MAX_RV	RV_NOT_DEFINED + 1

bool LOG_CONTAINS_USER_DATA[MAX_RV] = {
  false, // RVDK_NEWVOL
  false, // RVDK_FORMAT
  false, // RVDK_INITMAP
  ...
  true, // RVHF_INSERT
  true, // RVHF_DELETE
  true, // RVHF_UPDATE
  ...
};

Reply:
In my opinion, to determine which is faster, the cost of fewer ops but memory and cache load in the suggested code and more ops but no memory and cache load in the original code should be compared. However, I don't think it's worth it because it may not be the bottleneck.